### PR TITLE
Add docs and middleware for dynoid

### PR DESCRIPTION
**Pull request by kennyp**
Rationale
--------------
Dyno identity should be easy to adopt.

Changes
-------------
- Add docs and refactor `dynoid` and `dynoid/dynoidtest`.
- Add http middleware to `dynoid/middleware` to simplify incorporating dynoid.

Meta
--------
[Work Item](https://gus.lightning.force.com/a07EE00001mANddYAG)

### CODE REVIEW

File: dynoid/dynoid.go [hunk 1]
```diff
@@ -5,43 +5,234 @@ import (
...
 	"strings"
 	"sync"
 
 	"github.com/coreos/go-oidc/v3/oidc"
 )
 
-type IssuerCallback func(string) error
+const (
+	AudienceHeroku = "heroku"
+)
```
Comment (tt) [inline dynoid/dynoid.go:20]: Each service is supposed to use a unique audience so defining this as a constant might send the wrong signal.

File: dynoid/dynoid.go [hunk 2]
```diff
@@ -5,43 +5,234 @@ import (
...
 
-type Verifier struct {
-	IssuerCallback IssuerCallback
-	config         *oidc.Config
-	providers      sync.Map
+// Returned by an IssuerCallback get's an issuer it doesn't trust
```
Comment (tt) [inline dynoid/dynoid.go:22]:
```suggestion
// Returned by an IssuerCallback getting an issuer it doesn't trust
```

File: dynoid/dynoid.go [hunk 3]
```diff
@@ -5,43 +5,234 @@ import (
...
+	dyno := strings.Split(parts[4], ".")
+
+	if len(app) != 2 || len(dyno) != 2 {
+		return fmt.Errorf("unexpected subject format: %q", sub)
+	}
+
+	s.AppID = app[0]
+	s.AppName = app[1]
+	s.DynoName = dyno[0]
+	s.DynoNumber, _ = strconv.Atoi(dyno[1])
```
Comment (tt) [inline dynoid/dynoid.go:112]: This is a good example of where we need a later change due to [DR: Deprecating Dyno Numbering on the Heroku Platform](https://salesforce.quip.com/7teIAUvPUQiG) and [RFC: Fir Dyno Naming Separator](https://salesforce.quip.com/SiDbAIjz5w2i).

(Pinging @dianadoherty and @ricardochimal for their awareness.)

File: dynoid/dynoidtest/dynoidtest.go [hunk 4]
```diff
@@ -44,6 +59,10 @@ func (iss *Issuer) GenerateIDToken(clientID string) (string, error) {
 	return token.SignedString(iss.key)
 }
 
+func (iss *Issuer) Context() context.Context {
+	return iss.ctx
+}
+
```
Comment (tt) [inline dynoid/dynoidtest/dynoidtest.go:65]: This feels a bit dirty and seems to only be necessary for the test code. Can we pass a `context.Background()` through there to make the interface smaller?

File: hmiddleware/dynoid/dynoid.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+		token, err := verifier.VerifyHeroku(r.Context(), rawToken)
+
+		return r.WithContext(AddToContext(ctx, token, err))
+	}
+}
+
+func tokenFromHeader(r *http.Request) string {
+	bearer := r.Header.Get("Authorization")
+	if len(bearer) > 7 && strings.ToUpper(bearer[0:6]) == "BEARER" {
```
Comment (tt) [inline hmiddleware/dynoid/dynoid.go:117]: I suggest matching on the space too and making this case sensitive as recommended in [section 2.1 of RFC 7235](https://datatracker.ietf.org/doc/html/rfc7235#section-2.1):

> It uses a case-insensitive token as a means to identify the authentication scheme, followed by additional information necessary for achieving authentication via that scheme.

```suggestion
	if len(bearer) > 7 && bearer[:7] == "Bearer " {
```

**Review by tt — COMMENTED** [review]
This looks really good.

I do have one reservation about putting the middleware under the `hmiddleware` package even though it seems like the best fit. This module has quite a few dependencies and I think it would be great if we could eventually break it up a bit.

The dyno ID bits will need to remain together so may consider moving it to `dynoid/dynoidmiddleware` instead? That also prevents two packages that are named identically for import purposes.

File: dynoid/dynoid.go (hunk 3 above)
Comment (kennyp) [reply dynoid/dynoid.go:115]: I was thinking about this after watching @dianadoherty's VAT presentation this morning. Thanks for confirming...

What do you think about just returning `s.Dyno` and not parsing out the name from the number?

File: dynoid/dynoid.go (hunk 1 above)
Comment (kennyp) [reply dynoid/dynoid.go]: What if I just don't export the constant?

File: dynoid/dynoidtest/dynoidtest.go (hunk 4 above)
Comment (kennyp) [reply dynoid/dynoidtest/dynoidtest.go]: @tt I refactored it a bit to get rid of that. Let me know what you think.

File: dynoid/dynoid.go [hunk 6]
```diff
@@ -5,43 +5,230 @@ import (
...
+
+func (s *Subject) String() string {
+	if s == nil {
+		return ""
+	}
+
+	return fmt.Sprintf("app:%s.%s::dyno:%s", s.AppID, s.AppName, s.Dyno)
+}
+
+// Subject contains all of the token information stored by Heroku when issuing
```
Comment (mjrossi) [inline dynoid/dynoid.go:123]: Did you intend to rewrite this comment to Token specifically, or remove it?

**Review by mjrossi — APPROVED** [review]
I think a `readme.md` with some basics on how to use this package would be really helpful. That said, the inline comments definitely help, as does having the tests nearby to use as an example.

File: dynoid/dynoid.go (hunk 6 above)
Comment (kennyp) [reply dynoid/dynoid.go]: Good catch! Fixed.

File: dynoid/dynoid.go [hunk 7]
```diff
@@ -56,31 +241,60 @@ func (v *Verifier) Verify(ctx context.Context, rawIDToken string) (*oidc.IDToken
 	return verifier.Verify(ctx, rawIDToken)
 }
 
+// VerifyHeroku verifies the token and parses the returned issuer and subject
+// according to Heroku expected values
+func (v *Verifier) VerifyHeroku(ctx context.Context, rawIDToken string) (*Token, error) {
```
Comment (tt) [inline dynoid/dynoid.go:246]: Considering this is our own package, I think we should merge `Verify` and `VerifyHeroku`.

This seems simpler to maintain so we don't risk someone else starts using it for a use case it wasn't designed for.

File: dynoid/dynoid.go [hunk 8]
```diff
@@ -5,43 +5,228 @@ import (
...
 
-func New(clientID string) *Verifier {
-	return &Verifier{config: &oidc.Config{ClientID: clientID}}
+func (e *UntrustedIssuerError) Error() string {
+	return fmt.Sprintf("untrusted issuer: %v", e.Issuer)
+}
+
+// Returned when the token doesn't match the expected format
+type MalformedTokenError struct {
+	Err error
```
Comment (tt) [inline dynoid/dynoid.go:35]: We already define `Unwrap()` so this doesn't need to be exported:

```suggestion
	err error
```

File: dynoid/middleware/dynoid.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func internalServerError(format string, v ...any) func(http.Handler) http.Handler {
+	return func(http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, format, v...)
+		})
+	}
+}
```
Comment (tt) [inline dynoid/middleware/dynoid.go:122]: The formatting is unused:

```suggestion
func internalServerError(error string) func(http.Handler) http.Handler {
	return func(http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			http.Error(w, error, http.StatusInternalServerError)
		})
	}
}
```

File: dynoid/middleware/dynoid_test.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+func TestAuthorize(t *testing.T) {
+	router := chi.NewRouter()
+	router.Use(middleware.Authorize(
+		dynoidtest.Audience,
+		dynoid.AllowHerokuHost(dynoidtest.IssuerHost),
+	))
+	router.Get("/", func(w http.ResponseWriter, r *http.Request) {
+	})
```
Comment (tt) [inline dynoid/middleware/dynoid_test.go:25]: This is a nitpick but we don't need the chi dependency. Apologies if there's a syntax error...

```suggestion
	var handler http.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {})
	handler = middleware.Authorize(
		dynoidtest.Audience,
		dynoid.AllowHerokuHost(dynoidtest.IssuerHost),
	)(handler)
```

**Review by tt — APPROVED** [review]
A few suggestions but still looks good.

**Review by siraj-h — APPROVED** [review]

File: dynoid/dynoid.go (hunk 8 above)
Comment (kennyp) [reply dynoid/dynoid.go]: Updated in 1910815

File: dynoid/dynoid.go (hunk 7 above)
Comment (kennyp) [reply dynoid/dynoid.go]: Fixed in 91aad96

**Comment by kennyp** [comment]
> A few suggestions but still looks good.

Thanks for the suggestions @tt. I've incorporated your feedback.

**Comment by kennyp** [comment]
> I think a `readme.md` with some basics on how to use this package would be really helpful. That said, the inline comments definitely help, as does having the tests nearby to use as an example.

I agree with you @mjrossi, but I figured I would hold off on that until we some devcenter articles and a blog post to link to.

**Pull request merged**